datafusion/optimizer/src/push_down_limit.rs

```diff
@@ -153,6 +153,35 @@ impl OptimizerRule for PushDownLimit {
                 subquery_alias.input = Arc::new(new_limit);
                 Ok(Transformed::yes(LogicalPlan::SubqueryAlias(subquery_alias)))
             }
+            LogicalPlan::Extension(extension_plan)
+                if !extension_plan.node.allows_limit_to_inputs() =>
+            {
+                // If push down is not allowed, keep the original limit
+                original_limit(skip, fetch, LogicalPlan::Extension(extension_plan))
+            }
```

Comment:
I think technically you could remove this clause and it would be handled by the `input => ...` clause below

```suggestion

```